File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int ans = 0;

    int helper(int idx, int prev_idx,vector<int>& nums, vector<vector<int>>& dp) {
        if(idx == nums.size()){
            return 0;
        }
        if(dp[idx][prev_idx+1] != -1) return dp[idx][prev_idx+1];
        int pick =0;
        if(prev_idx == -1 || nums[idx] > nums[prev_idx]){
            pick = 1 + helper(idx+1,idx,nums,dp);
        }
        int notpick = helper(idx+1,prev_idx,nums,dp);

        return dp[idx][prev_idx+1] = max(pick,notpick);
    }

    int lengthOfLIS(vector<int>& nums) {
       vector<vector<int>>dp(nums.size(),vector<int>(nums.size(),-1));
       return helper(0,-1, nums, dp);
        
    }
};

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
   
   int help(int idx , int[] nums, int[][] dp,int previdx){
       if(idx == nums.length) return 0;

       if(dp[idx][previdx+1] != -1) return dp[idx][previdx+1];

       if(previdx == -1 || nums[previdx] < nums[idx]){
        return dp[idx][previdx+1] = Math.max(1+help(idx+1,nums,dp,idx),help(idx+1,nums,dp,previdx));
       }
       return dp[idx][previdx+1] = help(idx+1,nums,dp,previdx);
   }
    
    public int lengthOfLIS(int[] nums) {
       
         int[][] dp = new int[nums.length][nums.length+1];

         for(int[] a: dp){
          Arrays.fill(a,-1);
         }
         

          return help(0,nums,dp,-1);
        
          
    }
}